### ISSUE TITLE
想求一个用paddle、AlexNet模型做mnist训练的代码

### ISSUE THREAD

**Issue by jxydlj** [PaddlePaddle/Paddle#23719]
纯小白一个，学的很死，AI Studio上只有用LeNet做mnist的代码，尝试自己修改成用AlexNet模型但一直无法成功，望有大佬手写或分享一个paddle用AlexNet模型mnist的代码

**Comment by jxydlj** [comment]
import paddle
import paddle.fluid as fluid
import numpy as np
from paddle.fluid.dygraph.nn import Conv2D, Pool2D, Linear


# 定义 AlexNet 网络结构
class AlexNet(fluid.dygraph.Layer):
    def __init__(self, name_scope, num_classes=3):
        super(AlexNet, self).__init__(name_scope)

        # AlexNet与LeNet一样也会同时使用卷积和池化层提取图像特征
        # 与LeNet不同的是激活函数换成了‘relu’
        self.conv1 = Conv2D(num_channels=1, num_filters=96, filter_size=11, stride=4, padding=5, act='relu')
        self.pool1 = Pool2D(pool_size=2, pool_stride=2, pool_type='max')
        self.conv2 = Conv2D(num_channels=96, num_filters=256, filter_size=5, stride=1, padding=2, act='relu')
        self.pool2 = Pool2D(pool_size=2, pool_stride=2, pool_type='max')
        self.conv3 = Conv2D(num_channels=256, num_filters=384, filter_size=3, stride=1, padding=1, act='relu')
        self.conv4 = Conv2D(num_channels=384, num_filters=384, filter_size=3, stride=1, padding=1, act='relu')
        self.conv5 = Conv2D(num_channels=384, num_filters=256, filter_size=3, stride=1, padding=1, act='relu')
        self.pool5 = Pool2D(pool_size=2, pool_stride=2, pool_type='max')

        self.fc1 = Linear(input_dim=12544, output_dim=4096, act='relu')
        # self.fc1 = Linear(input_dim=256, output_dim=4096, act='relu')
        self.drop_ratio1 = 0.5
        self.fc2 = Linear(input_dim=4096, output_dim=4096, act='relu')
        self.drop_ratio2 = 0.5
        self.fc3 = Linear(input_dim=4096, output_dim=num_classes)

    def forward(self, x):
        x = self.conv1(x)
        x = self.pool1(x)
        x = self.conv2(x)
        x = self.pool2(x)
        x = self.conv3(x)
        x = self.conv4(x)
        x = self.conv5(x)
        x = self.pool5(x)
        x = fluid.layers.reshape(x, [x.shape[0], -1])
        x = self.fc1(x)
        # 在全连接之后使用dropout抑制过拟合
        x = fluid.layers.dropout(x, self.drop_ratio1)
        x = self.fc2(x)
        # 在全连接之后使用dropout抑制过拟合
        x = fluid.layers.dropout(x, self.drop_ratio2)
        x = self.fc3(x)
        return x



# 定义训练过程
def train(model):
    print('start training ... ')
    model.train()
    epoch_num = 5
    opt = fluid.optimizer.Momentum(learning_rate=0.001, momentum=0.9, parameter_list=model.parameters())
    # 使用Paddle自动下载数据并读取
    train_loader = paddle.batch(paddle.dataset.mnist.train(), batch_size=10)
    valid_loader = paddle.batch(paddle.dataset.mnist.test(), batch_size=10)
    for epoch in range(epoch_num):
        for batch_id, data in enumerate(train_loader()):
            # 调整输入数据形状和类型
            x_data = np.array([item[0] for item in data], dtype='float32').reshape(-1, 1, 28, 28)
            y_data = np.array([item[1] for item in data], dtype='int64').reshape(-1, 1)
            # 将numpy.ndarray转化成Tensor
            img = fluid.dygraph.to_variable(x_data)
            label = fluid.dygraph.to_variable(y_data)
            # 计算模型输出
            logits = model(img)
            # 计算损失函数
            loss = fluid.layers.softmax_with_cross_entropy(logits, label)
            avg_loss = fluid.layers.mean(loss)
            if batch_id % 1000 == 0:
                print("epoch: {}, batch_id: {}, loss is: {}".format(epoch, batch_id, avg_loss.numpy()))
            avg_loss.backward()
            opt.minimize(avg_loss)
            model.clear_gradients()

        model.eval()
        accuracies = []
        losses = []
        for batch_id, data in enumerate(valid_loader()):
            # 调整输入数据形状和类型
            x_data = np.array([item[0] for item in data], dtype='float32').reshape(-1, 1, 28, 28)
            y_data = np.array([item[1] for item in data], dtype='int64').reshape(-1, 1)
            # 将numpy.ndarray转化成Tensor
            img = fluid.dygraph.to_variable(x_data)
            label = fluid.dygraph.to_variable(y_data)
            # 计算模型输出
            logits = model(img)
            pred = fluid.layers.softmax(logits)
            # 计算损失函数
            loss = fluid.layers.softmax_with_cross_entropy(logits, label)
            acc = fluid.layers.accuracy(pred, label)
            accuracies.append(acc.numpy())
            losses.append(loss.numpy())
        print("[validation] accuracy/loss: {}/{}".format(np.mean(accuracies), np.mean(losses)))
        model.train()

    # 保存模型参数
    fluid.save_dygraph(model.state_dict(), 'mnist')


if __name__ == '__main__':
    # 创建模型
    with fluid.dygraph.guard():
        model = AlexNet("AlexNet", num_classes=10)
        #启动训练过程
        train(model)

**Comment by jxydlj** [comment]
望不吝赐教，指出问题

**Comment by liupluswei** [comment]
你好， 我们模型库中暂时还没有写好的动态图AlexNet 代码，请描述一下你碰到的具体问题，我这边可以找相关同事解答一下

**Comment by jxydlj** [comment]
> 你好， 我们模型库中暂时还没有写好的动态图AlexNet 代码，请描述一下你碰到的具体问题，我这边可以找相关同事解答一下

感谢您看到并回复了我

我是一直跟着AI Studio上的深度学习入门视频学习的，因学习时间较短水平较低，还处于搬运代码的水平，这个代码就是由LeNet做mnist的训练部分以及定义AlexNet模型部分拼接起来的。主要问题在于：
1、不清楚第一个全连接层中的输入数据是怎么来的，该如何计算这个数据（报错可能点之一）
2、训练过程中输入数据形状的转化可能有问题

如果可能的话占用您或其他一点时间帮我修改指正。

再次感谢

**Comment by lijianshe02** [comment]
> > 你好， 我们模型库中暂时还没有写好的动态图AlexNet 代码，请描述一下你碰到的具体问题，我这边可以找相关同事解答一下
> 
> 感谢您看到并回复了我
> 
> 我是一直跟着AI Studio上的深度学习入门视频学习的，因学习时间较短水平较低，还处于搬运代码的水平，这个代码就是由LeNet做mnist的训练部分以及定义AlexNet模型部分拼接起来的。主要问题在于：
> 1、不清楚第一个全连接层中的输入数据是怎么来的，该如何计算这个数据（报错可能点之一）
> 2、训练过程中输入数据形状的转化可能有问题
> 
> 如果可能的话占用您或其他一点时间帮我修改指正。
> 
> 再次感谢

请参考： https://github.com/PaddlePaddle/models/blob/develop/PaddleCV/image_classification/models/alexnet.py

**Comment by lijianshe02** [comment]
> > 你好， 我们模型库中暂时还没有写好的动态图AlexNet 代码，请描述一下你碰到的具体问题，我这边可以找相关同事解答一下
> 
> 感谢您看到并回复了我
> 
> 我是一直跟着AI Studio上的深度学习入门视频学习的，因学习时间较短水平较低，还处于搬运代码的水平，这个代码就是由LeNet做mnist的训练部分以及定义AlexNet模型部分拼接起来的。主要问题在于：
> 1、不清楚第一个全连接层中的输入数据是怎么来的，该如何计算这个数据（报错可能点之一）
> 2、训练过程中输入数据形状的转化可能有问题
> 
> 如果可能的话占用您或其他一点时间帮我修改指正。
> 
> 再次感谢

 feature在进入fc1层之前，需要进行reshape操作，假如reshape之前的feature shape为[N, C, H, W], 那么需要将该feature reshape成[N, C*H*W], 然后将reshape后的数据输入到fc1中；对应你的代码中这个：
    x = self.pool5(x)
    x = fluid.layers.reshape(x, [x.shape[0], -1])
    x = self.fc1(x)

**Comment by jxydlj** [comment]
> > > 你好， 我们模型库中暂时还没有写好的动态图AlexNet 代码，请描述一下你碰到的具体问题，我这边可以找相关同事解答一下
> > 
> > 
> > 感谢您看到并回复了我
> > 我是一直跟着AI Studio上的深度学习入门视频学习的，因学习时间较短水平较低，还处于搬运代码的水平，这个代码就是由LeNet做mnist的训练部分以及定义AlexNet模型部分拼接起来的。主要问题在于：
> > 1、不清楚第一个全连接层中的输入数据是怎么来的，该如何计算这个数据（报错可能点之一）
> > 2、训练过程中输入数据形状的转化可能有问题
> > 如果可能的话占用您或其他一点时间帮我修改指正。
> > 再次感谢
> 
> feature在进入fc1层之前，需要进行reshape操作，假如reshape之前的feature shape为[N, C, H, W], 那么需要将该feature reshape成[N, C_H_W], 然后将reshape后的数据输入到fc1中；对应你的代码中这个：
> x = self.pool5(x)
> x = fluid.layers.reshape(x, [x.shape[0], -1])
> x = self.fc1(x)

emmm，需要怎么改呢

**Comment by shippingwang** [comment]
> self.fc1 = Linear(input_dim=12544, output_dim=4096, act='relu')

你是说这个input_dim不确定size多少吧？可以print一下

>     x = self.pool5(x)

看看大小 之后再填
